### PR TITLE
ItemBHeart (Heart Container) OK and Documented

### PR DESCRIPTION
**Pull request by StickyThwomp**
Decompiled and documented the Heart Container.
- Updated Twinmold and Gyorg to use defined params.

### CODE REVIEW

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c [hunk 1]
```diff
@@ -10,26 +16,103 @@ void ItemBHeart_Destroy(Actor* thisx, PlayState* play);
...
+
+const ActorInit Item_B_Heart_InitVars = { ACTOR_ITEM_B_HEART,
+                                          ACTORCAT_BOSS,
+                                          FLAGS,
+                                          OBJECT_GI_HEARTS,
+                                          sizeof(ItemBHeart),
+                                          (ActorFunc)ItemBHeart_Init,
+                                          (ActorFunc)ItemBHeart_Destroy,
+                                          (ActorFunc)ItemBHeart_Update,
+                                          (ActorFunc)ItemBHeart_Draw };
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:29]: Could you add a trailing comma and reformat?
```suggestion
const ActorInit Item_B_Heart_InitVars = { ACTOR_ITEM_B_HEART,
                                          ACTORCAT_BOSS,
                                          FLAGS,
                                          OBJECT_GI_HEARTS,
                                          sizeof(ItemBHeart),
                                          (ActorFunc)ItemBHeart_Init,
                                          (ActorFunc)ItemBHeart_Destroy,
                                          (ActorFunc)ItemBHeart_Update,
                                          (ActorFunc)ItemBHeart_Draw, };
```

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c [hunk 2]
```diff
@@ -10,26 +16,103 @@ void ItemBHeart_Destroy(Actor* thisx, PlayState* play);
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Item_B_Heart/ItemBHeart_Init.s")
+void ItemBHeart_Init(Actor* thisx, PlayState* play) {
+    ItemBHeart* this = THIS;
+
+    if (Flags_GetCollectible(play, 0x1F)) {
+        Actor_Kill(&this->actor);
+    } else {
+        Actor_ProcessInitChain(&this->actor, sInitChain);
+        ActorShape_Init(&this->actor.shape, 0.0f, NULL, 0.8f);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:45]: We prefer to use early returns for Actor_Kill

```suggestion
    if (Flags_GetCollectible(play, 0x1F)) {
        Actor_Kill(&this->actor);
        return;
    }
    Actor_ProcessInitChain(&this->actor, sInitChain);
    ActorShape_Init(&this->actor.shape, 0.0f, NULL, 0.8f);
```

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c [hunk 3]
```diff
@@ -10,26 +16,103 @@ void ItemBHeart_Destroy(Actor* thisx, PlayState* play);
...
+
+/**
+ * Adjusts size and handles collection (if of proper scale)
+ */
+void ItemBHeart_Update(Actor* thisx, PlayState* play) {
+    ItemBHeart* this = THIS;
+
+    ItemBHeart_UpdateRotationAndScale(this, play);
+
+    if (!(this->scale < BHEART_SCALE_MIN_COLLECTIBLE)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:66]: Does this match?
```suggestion
    if (this->scale >= BHEART_SCALE_MIN_COLLECTIBLE) {
```

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c [hunk 4]
```diff
@@ -10,26 +16,103 @@ void ItemBHeart_Destroy(Actor* thisx, PlayState* play);
...
+        if (Actor_HasParent(&this->actor, play)) {
+            Flags_SetCollectible(play, 0x1F);
+            Actor_Kill(&this->actor);
+        } else {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:70]: Same early return comment

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c (hunk 3 above)
Comment (StickyThwomp) [reply src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:68]: It does not match.

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c [hunk 5]
```diff
@@ -20,16 +27,94 @@ const ActorInit Item_B_Heart_InitVars = {
...
+            return;
+        }
+        Actor_OfferGetItem(&this->actor, play, GI_HEART_CONTAINER, 30.0f, 80.0f);
+    }
+}
+
+/**
+ * Rotate continuously while approaching 40% of object's unit scale.
+ */
+void ItemBHeart_UpdateRotationAndScale(ItemBHeart* this, PlayState* play) {
```
Comment (hensldm) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:81]: Thoughts on maybe just `ItemBHeart_UpdateModel`? Feel like it still describes it pretty well, and is more succinct.

**Review by hensldm — APPROVED** [review]
1 thought, though it is probably fine as is so will approve.

**Review by AngheloAlf — APPROVED** [review]

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.h [hunk 6]
```diff
@@ -3,12 +3,20 @@
...
 #include "global.h"
 
-struct ItemBHeart;
+typedef enum {
+    /*   0 */ BHEART_PARAM_NORMAL,     // Spawn Parameter for Normal Scale.
+    /*  35 */ BHEART_PARAM_SMALL = 35, // Spawn Parameter for Small Scale.
+} ItemBHeartParams;
+
+#define BHEART_SCALE_SMALL (0.1f)           // Scale for Small Mode
+#define BHEART_SCALE_NORMAL (1.0f)          // Scale value for Normal Mode
```
Comment (engineer124) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.h:12]:
```suggestion
#define BHEART_SCALE_SMALL (0.1f)           // Scale for Small Mode
#define BHEART_SCALE_NORMAL (1.0f)          // Scale for Normal Mode
```
Or do `Scale value` for both

**Review by engineer124 — COMMENTED** [review]
Just two small nitpicks

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c [hunk 7]
```diff
@@ -20,16 +27,94 @@ const ActorInit Item_B_Heart_InitVars = {
...
+ * Rotate continuously while approaching 40% of object's unit scale.
+ */
+void ItemBHeart_UpdateModel(ItemBHeart* this, PlayState* play) {
+    this->actor.shape.rot.y += 0x400;
+    Math_ApproachF(&this->variableScale, 0.4f, 0.1f, 0.01f);
+    Actor_SetScale(&this->actor, this->variableScale * this->scale);
+}
+
+/**
+ * Draw translucently when infront of a boss warp portal
```
Comment (tom-overton) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:88]:
```suggestion
 * Draw translucently when in front of a boss warp portal
```

File: src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.h [hunk 8]
```diff
@@ -3,12 +3,20 @@
...
+#define BHEART_SCALE_MIN_COLLECTIBLE (0.5f) // unitScale <50% will disable collection
 
 typedef struct ItemBHeart {
     /* 0x000 */ Actor actor;
-    /* 0x144 */ UNK_TYPE1 unk_144[0x24];
-    /* 0x168 */ f32 unk_168;
+    /* 0x144 */ UNK_TYPE1 unk_144[0x20];
+    /* 0x164 */ f32 variableScale;
+    /* 0x168 */ f32 scale;
```
Comment (tom-overton) [inline src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.h:19]: What do you think about calling this `unitScale`? You used that in the comment above, and I like how it differentiates it from the *actor's* `scale` variable.

Comment (engineer124) [reply src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.h]: `unitScale` makes me think it always has to be `1.0f` (i.e. a unit value). What about `baseScale`?

Comment (hensldm) [reply src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.h]: Just want to bump on engineer's comment

Comment (StickyThwomp) [reply src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.h]: baseScale is fine, will rename. (TY again for the poke hensldm)

**Review by isghj5 — APPROVED** [review]

**Review by engineer124 — APPROVED** [review]

**Comment by engineer124** [comment]
Just need to merge master and run namefixer on the file for Jenkins to be happy

**Pull request merged**